### PR TITLE
Add support for duplicate txids.

### PR DESCRIPTION
**Pull request by karknu**
When the server asks for new txids from the client it can't assume that
those txids doesn't already exist in its buffer.

The TxSubmission test suite has been updated so that it doesn't generate
unique txids anymore.

### CODE REVIEW

**Comment by karknu** [comment]
Fixes #2680 and #2682.

File: ouroboros-network/src/Ouroboros/Network/TxSubmission/Inbound.hs
Comment (mrBliss) [inline ouroboros-network/src/Ouroboros/Network/TxSubmission/Inbound.hs:328]:
```suggestion
            !bufferedTxs''' = bufferedTxs'' <> Map.fromList (zip live (repeat Nothing))
```
Also, the moment we need three primes (`'''`) might be a sign that we should switch to a different naming scheme. Either by numbering or by using different, more descriptive names.

File: ouroboros-network/src/Ouroboros/Network/TxSubmission/Inbound.hs
Comment (mrBliss) [inline ouroboros-network/src/Ouroboros/Network/TxSubmission/Inbound.hs:346]:
```suggestion
                              -> StrictSeq txid
                              -> Map txid TxSizeInBytes
```

File: ouroboros-network/src/Ouroboros/Network/TxSubmission/Inbound.hs
Comment (mrBliss) [inline ouroboros-network/src/Ouroboros/Network/TxSubmission/Inbound.hs:385]:
```suggestion
```

File: ouroboros-network/src/Ouroboros/Network/TxSubmission/Inbound.hs
Comment (mrBliss) [inline ouroboros-network/src/Ouroboros/Network/TxSubmission/Inbound.hs:398]:
```suggestion
        -- are not still in unacknowledgedTxIds''. This happens in case of duplicate
```

**Review by mrBliss — COMMENTED** [review]
What's the status of this PR?

File: ouroboros-network/src/Ouroboros/Network/TxSubmission/Inbound.hs
Comment (coot) [reply ouroboros-network/src/Ouroboros/Network/TxSubmission/Inbound.hs:328]: I am more and more leaning toward using unicode `bufferedTxs₃` (this makes `agda` code easier to read - though harder to type).

**Comment by karknu** [comment]
> What's the status of this PR?

From my point of view it is ready and waiting on reviews.

File: ouroboros-network/src/Ouroboros/Network/TxSubmission/Inbound.hs
Comment (karknu) [reply ouroboros-network/src/Ouroboros/Network/TxSubmission/Inbound.hs:328]: I'll figure out a better naming scheme that doesn't require unicode.

**Comment by mrBliss** [comment]
@nfrisby can you take a look at this?

File: ouroboros-network/src/Ouroboros/Network/TxSubmission/Inbound.hs
Comment (coot) [reply ouroboros-network/src/Ouroboros/Network/TxSubmission/Inbound.hs:328]: The revolution is still to happen :D ! Unicode has its WATs: how's that possible that there's not supscript `b,c,d`.

**Comment by coot** [comment]
Rebasing on top of `mater`  should fix `Windows` CI build error.

File: ouroboros-network/src/Ouroboros/Network/TxSubmission/Inbound.hs
Comment (mrBliss) [inline ouroboros-network/src/Ouroboros/Network/TxSubmission/Inbound.hs:326]: Does this `forceElemsToWHNF` force the thunks? Because the type of `bufferedTxs` is `Map txid (Maybe tx)`, so even forcing a `Maybe tx` to WHNF won't force the `tx` to WHNF (unless it already is).

What we in consensus typically do is replace `Maybe` with a strict variant, often specific to the use case. The constructor names would then also explain what `Nothing` and `Just` mean.

Actually, could you try removing the `forceElemsToWHNF` and replace `<>` with `Map.union`? I think the problem might actually be that `Map.Strict` is the same type as `Map.Lazy` and that the `Semigroup` instance of `Map.Lazy` is being used here.

Comment (karknu) [reply ouroboros-network/src/Ouroboros/Network/TxSubmission/Inbound.hs:326]: Removing `forceElemsToWHNF` and replacing `<>` with `Map.union` seems to still work. However I seem to be unable to trigger a "thunk invariance failure" at all so I'm not going to address this.

File: ouroboros-network/src/Ouroboros/Network/TxSubmission/Inbound.hs
Comment (nfrisby) [inline ouroboros-network/src/Ouroboros/Network/TxSubmission/Inbound.hs:324]: > duplicate txids within the same window.

Is there a simple-enough example you could record in this code comment of how this might happen? I don't recall; it's been too long since I touched that repro we had a while ago.

File: ouroboros-network/src/Ouroboros/Network/TxSubmission/Inbound.hs
Comment (nfrisby) [inline ouroboros-network/src/Ouroboros/Network/TxSubmission/Inbound.hs:364]: Use `Map.filterWithKey`?

File: ouroboros-network/src/Ouroboros/Network/TxSubmission/Inbound.hs
Comment (nfrisby) [inline ouroboros-network/src/Ouroboros/Network/TxSubmission/Inbound.hs:396]: Worth creating a set for use when checking membership? As I mentioned elsewhere, I don't have a sense of what the expected sizes will be.

**Review by nfrisby — COMMENTED** [review]
This is tricky stuff! I'm glad you figured out the bug.

I apologize for only responding two weeks later now :(

I left a bunch of comments throughout, some of which are minor. Mostly about code clarity, suggestive names, etc.

The potentially blockers are the questions about efficiency (some expressions went from logarithmic to linear eg) but I don't know the relative sizes involved. Have you considered those trade-offs? If so, it'd be valuable to record your thoughts in the comments.

File: ouroboros-network/src/Ouroboros/Network/TxSubmission/Outbound.hs
Comment (karknu) [reply ouroboros-network/src/Ouroboros/Network/TxSubmission/Outbound.hs:177]: `unackedMap` has to be removed since it is keyed on `txid` and can't cope with duplicate txid in the same window.
The tx window size is 10. It is not configurable and requires a new protocol version to change.

File: ouroboros-network/test/Test/Ouroboros/Network/TxSubmission.hs
Comment (karknu) [reply ouroboros-network/test/Test/Ouroboros/Network/TxSubmission.hs:297]: I prefer to keep it.

File: ouroboros-network/src/Ouroboros/Network/TxSubmission/Inbound.hs
Comment (karknu) [reply ouroboros-network/src/Ouroboros/Network/TxSubmission/Inbound.hs:396]: The tx windows size is 10 so I don't think a Set is worth it.

File: ouroboros-network/src/Ouroboros/Network/TxSubmission/Inbound.hs
Comment (karknu) [reply ouroboros-network/src/Ouroboros/Network/TxSubmission/Inbound.hs:268]: Perhaps `acknowledgeTxIds` is more suitable. The function acknowledges txids that are either in mempool or that we've learned of already but not yet comited.

File: ouroboros-network/src/Ouroboros/Network/TxSubmission/Inbound.hs
Comment (karknu) [reply ouroboros-network/src/Ouroboros/Network/TxSubmission/Inbound.hs:317]: Will change to `bufferedTxsN`.

File: ouroboros-network/src/Ouroboros/Network/TxSubmission/Inbound.hs
Comment (karknu) [reply ouroboros-network/src/Ouroboros/Network/TxSubmission/Inbound.hs:324]: This is the simplest example that quickcheck could generate. I don't think it is concise enough to put it in a code comment.
Consider the following list of TXs and a window size greater than or equal to five.
[Tx 0, Tx 1, Tx 2, Tx 3, Tx 2]

 1. The server requests 3 new TxIds.
 2. The client responds with [TxId 0, TxId 1, TxId 2].
 3. The server requests [Tx 0, Tx 1].
 4. The server acks [TxId 0, TxId 1] and request 2 new TxIds.
 5. The client responds with [Tx 0, Tx 1]
 6. The server reqest Tx 2.
 7. The client responds with [TxId 3, TxId 2 (dupe)].
 8. The client responds with Tx 2.
 9. The server request Tx 3
10. The server acks [TxId 2, TxId 3] and request two new TxIds. Note that only the TxId 2 from msg 2 is acked, not the dupe TxId2 from msg 7.
11. The client responds with no more Txids.
12. The server will trigger an assertion failure (line 199) because it has failed to acknowledge the 2 (dupe) TxId.

File: ouroboros-network/src/Ouroboros/Network/TxSubmission/Inbound.hs
Comment (karknu) [reply ouroboros-network/src/Ouroboros/Network/TxSubmission/Inbound.hs:325]: Will fix.

File: ouroboros-network/src/Ouroboros/Network/TxSubmission/Inbound.hs
Comment (karknu) [reply ouroboros-network/src/Ouroboros/Network/TxSubmission/Inbound.hs:364]: Will change.

File: ouroboros-network/test/Test/Ouroboros/Network/TxSubmission.hs
Comment (karknu) [reply ouroboros-network/test/Test/Ouroboros/Network/TxSubmission.hs:322]: I have no interest in including `go` directly into my property. I prefer to keep it in a form that makes it easy to cut-and-paste into every future test case I write.

**Comment by karknu** [comment]
> This is tricky stuff! I'm glad you figured out the bug.
> 
> I apologize for only responding two weeks later now :(
> 
> I left a bunch of comments throughout, some of which are minor. Mostly about code clarity, suggestive names, etc.
> 
> The potentially blockers are the questions about efficiency (some expressions went from logarithmic to linear eg) but I don't know the relative sizes involved. Have you considered those trade-offs? If so, it'd be valuable to record your thoughts in the comments.

Thank you for your review. The current window size is 10, so I have not considered complexity.
The most obvious performance improvement would be to fetch requests in order (#1721). The arbitrary fetching order increases the time from when we receive a new Tx to when we can commit it to the mempool by several RTTs.
There are other TxSubmission related issues that I hope to get to address soon such as #2688, #2615, and #1736.
However this PR is only for duplicate TX id support.

File: ouroboros-network/src/Ouroboros/Network/TxSubmission/Inbound.hs
Comment (nfrisby) [reply ouroboros-network/src/Ouroboros/Network/TxSubmission/Inbound.hs:324]: Thanks, that's helpful.

Hmm. Could the comment say "duplicate txids can arise merely as a consequence of pipelining the current BF protocol"? Do you think that is an accurate summary?

File: ouroboros-network/src/Ouroboros/Network/TxSubmission/Inbound.hs
Comment (nfrisby) [reply ouroboros-network/src/Ouroboros/Network/TxSubmission/Inbound.hs:268]: `acknowledgeTxIds` sounds good to me 👍. Having recently revisited the documentation on the state fields, the fact that this involves updating `bufferedTxs` now seems unsurprising to me.

File: ouroboros-network/test/Test/Ouroboros/Network/TxSubmission.hs
Comment (nfrisby) [reply ouroboros-network/test/Test/Ouroboros/Network/TxSubmission.hs:322]: With that feedback, I'll refine my suggestion to define a combinator like this one. The main refinement is that it doesn't necessitate the `[String]` space-leak and so seems like a good candidate for inclusion in `io-sim`.

```
foldSayResultTrace ::
     (Time -> ThreadId -> String -> a -> a)   -- ^ handle an 'EventSay'
  -> (a -> Either Failure b -> c)  -- ^ handle end of trace
  -> (a -> SomeException -> c)  -- ^ handle pure exception thrown by definition of trace
  -> a
  -> Trace b
  -> c
```

Main point: Whatever the details end up being, I agree that a special combinator focusing only on `say` and termination events would be helpful. It could be another `TraceSay` data type with its own fold function or just the Church encoding as above, no personal preference. For example, I'm mainly envisioning `a ~ c ~ Test.QuickCheck.Property`.

**Comment by nfrisby** [comment]
> The current window size is 10, so I have not considered complexity.

I think "the window size is small" resolves all my complexity concerns. I'm tempted to record that in a comment as relevant contextual information. Your call 👍.

Last question: do you want to factor out the `io-sim` wrapper before merging this PR?

File: ouroboros-network/src/Ouroboros/Network/TxSubmission/Inbound.hs
Comment (karknu) [reply ouroboros-network/src/Ouroboros/Network/TxSubmission/Inbound.hs:324]: In the example above pipelining permitted the duplicate txids to trigger an assertion, but pipelining didn't create the duplicate Tx id. The client decided to send a duplicate Tx id within the same window. I think in theory this could happen in case the client sent us a new TxId, rolled back the a fork without that Tx, fetched the Tx again from another peer and then sent us the same TxId again.

It is also possible to trigger this bug if the Txs aren't fetched in order (they are fetched based on `Ord` TxId, not in the order they where presented), in away that doesn't depend on pipelining.

**Comment by karknu** [comment]
> > The current window size is 10, so I have not considered complexity.
> 
> I think "the window size is small" resolves all my complexity concerns. I'm tempted to record that in a comment as relevant contextual information. Your call 👍.

I added a comment regarding the windows size.

> 
> Last question: do you want to factor out the `io-sim` wrapper before merging this PR?

I expect more resistance/comments on adding `evaluateTrace` to io-sim so I'm fine with just having it as a TODO.

File: ouroboros-network/src/Ouroboros/Network/TxSubmission/Inbound.hs
Comment (mrBliss) [reply ouroboros-network/src/Ouroboros/Network/TxSubmission/Inbound.hs]: I still think that `Map.union` is cleaner than `forceElemsToWHNF` (no need to force thunks if you don't even introduce them).

**Review by nfrisby — APPROVED** [review]
OK, yeah, a max window size of 10 seems like it'll truncate any of these potential algorithmic complexity issues. Thanks! LGTM! Approving.

**Comment by mrBliss** [comment]
Any reason not to bors this?

**Comment by karknu** [comment]
> Any reason not to bors this?

I'm waiting for either @dcoutts or @coot . Hoping for 2021.

File: ouroboros-network/src/Ouroboros/Network/TxSubmission/Inbound.hs
Comment (coot) [inline ouroboros-network/src/Ouroboros/Network/TxSubmission/Inbound.hs:370]: This comment seems out of place now, I think it ought to be at the top of this hunk.

File: ouroboros-network/src/Ouroboros/Network/TxSubmission/Inbound.hs
Comment (coot) [inline ouroboros-network/src/Ouroboros/Network/TxSubmission/Inbound.hs:366]: One could use `notElem` here.

File: ouroboros-network/src/Ouroboros/Network/TxSubmission/Inbound.hs
Comment (karknu) [reply ouroboros-network/src/Ouroboros/Network/TxSubmission/Inbound.hs:380]: Yes, `bufferedTxs` contains txs that are inflight. We don't want them overwritten with `Nothing` in case of a duplicate txid.

File: ouroboros-network/src/Ouroboros/Network/TxSubmission/Inbound.hs
Comment (karknu) [reply ouroboros-network/src/Ouroboros/Network/TxSubmission/Inbound.hs:394]: `are still not` doesn't have the same meaning as `are not still`.

File: ouroboros-network/src/Ouroboros/Network/TxSubmission/Inbound.hs
Comment (karknu) [reply ouroboros-network/src/Ouroboros/Network/TxSubmission/Inbound.hs:326]: Will move it to `bufferedTxs2`.

File: ouroboros-network/src/Ouroboros/Network/TxSubmission/Inbound.hs
Comment (coot) [reply ouroboros-network/src/Ouroboros/Network/TxSubmission/Inbound.hs:394]: I see, that's ok then.

**Pull request merged**
